[SPIKE] Check all pages' accessibility

domoscargin opened the pull request:
Popped this together to check how long it'd take to run accessibility tests on ALL our pages. Just a hacky function to get all directories with an `index.html` file, then running an axe validation test on each of those.

Takes around 80 seconds on my machine, but open for ways to speed it up.

There's a few failures flagged, mostly to do with our heading order.

If taken forward, the code would need to be cleaned up and optimised.

Edit: looks like the tests took about 2m 45s.

domoscargin commented:
Had a brief look at `it.concurrent.each`, but at first glance this doesn't play nicely with Puppeteer.

colinrotherham committed:
Add Axe rules/reporting from GOV.UK Frontend

See: https://github.com/alphagov/govuk-frontend/commit/ebb9e4e11485bd39bf6d0141c445f1f29039af3e

colinrotherham committed:
Gather accessibility test URLs using `glob`

colinrotherham committed:
Group Axe excludes into array

colinrotherham commented:
@domoscargin Just rebased this to see if the [`@axe-core/puppeteer` update](https://github.com/alphagov/govuk-design-system/pull/3224) fixed the disconnect issue

Looking good locally

domoscargin commented:
Darn, failed here.

colinrotherham committed:
Run each accessibility test in a new page

colinrotherham committed:
Add Puppeteer options from GOV.UK Frontend

colinrotherham commented:
Another [Puppeteer update](https://github.com/alphagov/govuk-design-system/pull/3273) so rebasing this again

colinrotherham commented:
Had some more Axe and Puppeteer updates in https://github.com/alphagov/govuk-design-system/pull/3468 so giving this one a rebase

colinrotherham commented:
@domoscargin Looks like it's all working 🙌

Status check failures are now accessibility report issues, as intended

domoscargin commented:
# Failures notes

## `aria-allowed-role`
This flags our use of `role="region"` on the `<video>` element, which doesn't allow that role.

This was [done on purpose](https://github.com/alphagov/govuk-design-system/commit/686d1f16d11f2e0cb37566a50f698636da472258), but unclear if it's still relevant.

owenatgov commented:
I'm looking at this as part of https://github.com/alphagov/design-system-team-internal/issues/912 (internal repo) as it helps us with auditing the site for at the very least issues we've had reported to us on heading structure. I'm going to rebase it so that this is repeatable and referable but as part of this I'm going to drop https://github.com/alphagov/govuk-design-system/pull/2957/commits/21617841f61691299fef01c277e13e22c719e979 as this is where the majorty of conflicts are being reported from and we resolved this as part of https://github.com/alphagov/govuk-design-system/pull/4054

As a personal take, whilst we were nervous about how long this took to run per test run, I think it'd be valuable to just have and run manually as an ocassional auditing tool.

selfthinker commented:
I just looked at the failures.

`target-size`: That behaviour is indeed weird. I can confirm none of those failures exist and are not flagged by the Axe browser extension. I wonder if this check is simply buggy. It might be worth checking once a year or so if it's been fixed and enabling the rule again when it was.

`color-contrast-enhanced`: I agree this check makes sense to disable.

`aria-allowed-attr`: This seems generally like a sensible check. The browser extension does flag it for me as well. Is there a way to only disable it for certain components? It would be a shame to lose the check for other components than conditionally revealing radios.

`region`: Is "All page content should be contained by landmarks" the only thing this one checks? If it does, I agree we should disable it. If not, same with the above, would it be possible to only disable it for certain components? I see on the list of exceptions that certain areas like `.app-phase-banner` and `.govuk-skip-link` are also excluded. Similarly here, it would be a shame to lose those being checked as a whole just for this single reason.

`aria-allowed-role`: Changing the video role to a `region` is quite iffy, so definitely a good find by the tool. Although I can see from the relevant PR that there was a good amount of testing, even if this does not do anything weird at the moment, chances that it might do in the future are quite high. I would suggest to solve this differently.

selfthinker commented:
> `aria-allowed-role`: Changing the video role to a `region` is quite iffy, so definitely a good find by the tool. Although I can see from the relevant PR that there was a good amount of testing, even if this does not do anything weird at the moment, chances that it might do in the future are quite high. I would suggest to solve this differently.

@owenatgov reminded me today to add any potential solution. Here is what I would do:
Don't add the `role=region` and `aria-labelledby` to the `<video>` element but to the `<div>` surrounding it.

domoscargin committed:
Check all pages accessibility

domoscargin committed:
Split out excludes, and globs

It might be an axe bug, but exclude doesn't seem to work properly with an array argument. Splitting them out reduces failures from 300+ to 139.

Additionally, I've made some changes to which files are included in the glob, with a view to splitting these out and only running certain tests in certain environments to avoid jammingthings up for too long.

domoscargin committed:
Add pattern indexes

domoscargin committed:
[WIP] Fix accessibility failures

domoscargin committed:
Remove unneeded exclusion

domoscargin committed:
Fix aria-allowed-role failure

This was done on purpose in https://github.com/alphagov/govuk-design-system/pull/304, but unclear if it still applies?

domoscargin committed:
Remove phase banner exclude

This has been fixed by our work rearranging the template

domoscargin committed:
Remove unecessary exclusions

These have been fixed by our page template changes

domoscargin committed:
Enable target-size

We were getting some false flags here, because Axe was firing before the font had loaded, so the heights of some bits of text weren't 24px

For now, wait until networkidle0 and see if that works.
